File: ainpc-core-plugin/src/main/java/ro/ainpc/story/StoryContextSnapshot.java
package ro.ainpc.story;

import ro.ainpc.world.WorldContextSnapshot;
import ro.ainpc.world.WorldPlaceInfo;
import ro.ainpc.world.WorldRegionInfo;

import java.util.List;
import java.util.Locale;

public record StoryContextSnapshot(
    String subjectNpcName,
    String subjectNpcOccupation,
    String playerName,
    WorldContextSnapshot worldContext,
    RegionStoryState persistentRegionState,
    PlaceStoryState persistentPlaceState,
    List<StoryEvent> recentStoryEvents,
    List<QuestAnchorSnapshot> activeQuestAnchors,
    List<String> storySignals,
    List<String> warnings
) {
    public StoryContextSnapshot {
        subjectNpcName = valueOrEmpty(subjectNpcName);
        subjectNpcOccupation = valueOrEmpty(subjectNpcOccupation);
        playerName = valueOrEmpty(playerName);
        worldContext = worldContext != null ? worldContext : WorldContextSnapshot.empty();
        recentStoryEvents = List.copyOf(recentStoryEvents != null ? recentStoryEvents : List.of());
        activeQuestAnchors = List.copyOf(activeQuestAnchors != null ? activeQuestAnchors : List.of());
        storySignals = List.copyOf(storySignals != null ? storySignals : List.of());
        warnings = List.copyOf(warnings != null ? warnings : List.of());
    }

    public static StoryContextSnapshot empty() {
        return new StoryContextSnapshot(
            "",
            "",
            "",
            WorldContextSnapshot.empty(),
            null,
            null,
            List.of(),
            List.of(),
            List.of(),
            List.of()
        );
    }

    public boolean isEmpty() {
        return subjectNpcName.isBlank()
            && playerName.isBlank()
            && worldContext.isEmpty()
            && persistentRegionState == null
            && persistentPlaceState == null
            && recentStoryEvents.isEmpty()
            && activeQuestAnchors.isEmpty()
            && storySignals.isEmpty()
            && warnings.isEmpty();
    }

    public String toPromptBlock() {
        if (isEmpty()) {
            return "";
        }

        StringBuilder block = new StringBuilder("STORY_CONTEXT:\n");
        if (!subjectNpcName.isBlank()) {
            block.append("- subject_npc: ")
                .append(subjectNpcName)
                .append(", occupation=")
                .append(valueOrUnknown(subjectNpcOccupation))
                .append("\n");
        }
        if (!playerName.isBlank()) {
            block.append("- player: ").append(playerName).append("\n");
        }

        WorldRegionInfo region = worldContext.currentRegion();
        if (region != null) {
            block.append("- region_story: ")
                .append(region.id())
                .append(", state=")
                .append(valueOrUnknown(region.storyStateKey()))
                .append(", mode=")
                .append(region.storyMode().name().toLowerCase(Locale.ROOT))
                .append(", pool=")
                .append(region.storyPool())
                .append("\n");
        }

        if (persistentRegionState != null) {
            block.append("- persistent_region_story: ")
                .append(persistentRegionState.regionId())
                .append(", state=")
                .append(valueOrUnknown(persistentRegionState.stateKey()))
                .append(", mode=")
                .append(persistentRegionState.storyMode().name().toLowerCase(Locale.ROOT))
                .append(", pool=")
                .append(persistentRegionState.storyPool())
                .append(", vars=")
                .append(persistentRegionState.variables())
                .append("\n");
        }

        WorldPlaceInfo place = worldContext.currentPlace();
        if (place != null) {
            block.append("- place_story: ")
                .append(place.id())
                .append(", type=")
                .append(place.placeType().name().toLowerCase(Locale.ROOT))
                .append(", tags=")
                .append(place.tags())
                .append(", metadata=")
                .append(place.metadata())
                .append("\n");
        }

        if (persistentPlaceState != null) {
            block.append("- persistent_place_story: ")
                .append(persistentPlaceState.placeId())
                .append(", state=")
                .append(valueOrUnknown(persistentPlaceState.stateKey()))
                .append(", vars=")
                .append(persistentPlaceState.variables())
                .append("\n");
        }

        if (!recentStoryEvents.isEmpty()) {
            block.append("- recent_story_events:\n");
            for (StoryEvent event : recentStoryEvents) {
                block.append("  - ")
                    .append(event.scopeType())
                    .append(":")
                    .append(event.scopeId())
                    .append(" ")
                    .append(event.eventType())
                    .append("/")
                    .append(valueOrUnknown(event.eventKey()));
                if (!event.title().isBlank()) {
                    block.append(" - ").append(event.title());
                }
                block.append("\n");
            }
        }

        if (!storySignals.isEmpty()) {
            block.append("- story_signals: ").append(storySignals).append("\n");
        }

        if (!activeQuestAnchors.isEmpty()) {
            block.append("- active_quest_anchors:\n");
            for (QuestAnchorSnapshot anchor : activeQuestAnchors) {
                block.append("  - ")
                    .append(anchor.templateId())
                    .append(" ")
                    .append(anchor.objectiveKey())
                    .append(" [")
                    .append(anchor.objectiveType())
                    .append("] -> ")
                    .append(anchor.anchorType())
                    .append(":")
                    .append(anchor.anchorId())
                    .append(", status=")
                    .append(valueOrUnknown(anchor.questStatus()))
                    .append("\n");
            }
        }

        if (!warnings.isEmpty()) {
            block.append("- warnings: ").append(warnings).append("\n");
        }
        return block.toString();
    }

    private static String valueOrEmpty(String value) {
        return value == null ? "" : value;
    }

    private static String valueOrUnknown(String value) {
        return value == null || value.isBlank() ? "unknown" : value;
    }

    public record QuestAnchorSnapshot(
        String templateId,
        String questCode,
        String questStatus,
        String objectiveKey,
        String objectiveType,
        String reference,
        String anchorType,
        String anchorId,
        String anchorLabel,
        long updatedAt
    ) {
        public QuestAnchorSnapshot {
            templateId = valueOrEmpty(templateId);
            questCode = valueOrEmpty(questCode);
            questStatus = valueOrEmpty(questStatus);
            objectiveKey = valueOrEmpty(objectiveKey);
            objectiveType = valueOrEmpty(objectiveType);
            reference = valueOrEmpty(reference);
            anchorType = valueOrEmpty(anchorType);
            anchorId = valueOrEmpty(anchorId);
            anchorLabel = valueOrEmpty(anchorLabel);
        }
    }
}

File: ainpc-core-plugin/bin/main/ro/ainpc/story/StoryContextSnapshot.kt
package ro.ainpc.story

import ro.ainpc.world.WorldContextSnapshot
import ro.ainpc.world.WorldPlaceInfo
import ro.ainpc.world.WorldRegionInfo
import java.util.Locale

class StoryContextSnapshot(
    subjectNpcName: String?,
    subjectNpcOccupation: String?,
    playerName: String?,
    worldContext: WorldContextSnapshot?,
    private val persistentRegionStateValue: RegionStoryState?,
    private val persistentPlaceStateValue: PlaceStoryState?,
    recentStoryEvents: List<StoryEvent>?,
    activeQuestAnchors: List<QuestAnchorSnapshot>?,
    storySignals: List<String>?,
    warnings: List<String>?
) {
    private val subjectNpcNameValue: String = valueOrEmpty(subjectNpcName)
    private val subjectNpcOccupationValue: String = valueOrEmpty(subjectNpcOccupation)
    private val playerNameValue: String = valueOrEmpty(playerName)
    private val worldContextValue: WorldContextSnapshot = worldContext ?: WorldContextSnapshot.empty()
    private val recentStoryEventsValue: List<StoryEvent> = (recentStoryEvents ?: emptyList()).toList()
    private val activeQuestAnchorsValue: List<QuestAnchorSnapshot> = (activeQuestAnchors ?: emptyList()).toList()
    private val storySignalsValue: List<String> = (storySignals ?: emptyList()).toList()
    private val warningsValue: List<String> = (warnings ?: emptyList()).toList()

    fun subjectNpcName(): String = subjectNpcNameValue
    fun subjectNpcOccupation(): String = subjectNpcOccupationValue
    fun playerName(): String = playerNameValue
    fun worldContext(): WorldContextSnapshot = worldContextValue
    fun persistentRegionState(): RegionStoryState? = persistentRegionStateValue
    fun persistentPlaceState(): PlaceStoryState? = persistentPlaceStateValue
    fun recentStoryEvents(): List<StoryEvent> = recentStoryEventsValue
    fun activeQuestAnchors(): List<QuestAnchorSnapshot> = activeQuestAnchorsValue
    fun storySignals(): List<String> = storySignalsValue
    fun warnings(): List<String> = warningsValue

    fun isEmpty(): Boolean {
        return subjectNpcNameValue.isBlank() &&
            playerNameValue.isBlank() &&
            worldContextValue.isEmpty() &&
            persistentRegionStateValue == null &&
            persistentPlaceStateValue == null &&
            recentStoryEventsValue.isEmpty() &&
            activeQuestAnchorsValue.isEmpty() &&
            storySignalsValue.isEmpty() &&
            warningsValue.isEmpty()
    }

    fun toPromptBlock(): String {
        if (isEmpty()) {
            return ""
        }

        val block = StringBuilder("STORY_CONTEXT:\n")
        if (subjectNpcNameValue.isNotBlank()) {
            block.append("- subject_npc: ")
                .append(subjectNpcNameValue)
                .append(", occupation=")
                .append(valueOrUnknown(subjectNpcOccupationValue))
                .append("\n")
        }
        if (playerNameValue.isNotBlank()) {
            block.append("- player: ").append(playerNameValue).append("\n")
        }

        val region: WorldRegionInfo? = worldContextValue.currentRegion()
        if (region != null) {
            block.append("- region_story: ")
                .append(region.id())
                .append(", state=")
                .append(valueOrUnknown(region.storyStateKey()))
                .append(", mode=")
                .append(region.storyMode().name.lowercase(Locale.ROOT))
                .append(", pool=")
                .append(region.storyPool())
                .append("\n")
        }

        if (persistentRegionStateValue != null) {
            block.append("- persistent_region_story: ")
                .append(persistentRegionStateValue.regionId())
                .append(", state=")
                .append(valueOrUnknown(persistentRegionStateValue.stateKey()))
                .append(", mode=")
                .append(persistentRegionStateValue.storyMode().name.lowercase(Locale.ROOT))
                .append(", pool=")
                .append(persistentRegionStateValue.storyPool())
                .append(", vars=")
                .append(persistentRegionStateValue.variables())
                .append("\n")
        }

        val place: WorldPlaceInfo? = worldContextValue.currentPlace()
        if (place != null) {
            block.append("- place_story: ")
                .append(place.id())
                .append(", type=")
                .append(place.placeType().name.lowercase(Locale.ROOT))
                .append(", tags=")
                .append(place.tags())
                .append(", metadata=")
                .append(place.metadata())
                .append("\n")
        }

        if (persistentPlaceStateValue != null) {
            block.append("- persistent_place_story: ")
                .append(persistentPlaceStateValue.placeId())
                .append(", state=")
                .append(valueOrUnknown(persistentPlaceStateValue.stateKey()))
                .append(", vars=")
                .append(persistentPlaceStateValue.variables())
                .append("\n")
        }

        if (recentStoryEventsValue.isNotEmpty()) {
            block.append("- recent_story_events:\n")
            for (event in recentStoryEventsValue) {
                block.append("  - ")
                    .append(event.scopeType())
                    .append(":")
                    .append(event.scopeId())
                    .append(" ")
                    .append(event.eventType())
                    .append("/")
                    .append(valueOrUnknown(event.eventKey()))
                if (event.title().isNotBlank()) {
                    block.append(" - ").append(event.title())
                }
                block.append("\n")
            }
        }

        if (storySignalsValue.isNotEmpty()) {
            block.append("- story_signals: ").append(storySignalsValue).append("\n")
        }

        if (activeQuestAnchorsValue.isNotEmpty()) {
            block.append("- active_quest_anchors:\n")
            for (anchor in activeQuestAnchorsValue) {
                block.append("  - ")
                    .append(anchor.templateId())
                    .append(" ")
                    .append(anchor.objectiveKey())
                    .append(" [")
                    .append(anchor.objectiveType())
                    .append("] -> ")
                    .append(anchor.anchorType())
                    .append(":")
                    .append(anchor.anchorId())
                    .append(", status=")
                    .append(valueOrUnknown(anchor.questStatus()))
                    .append("\n")
            }
        }

        if (warningsValue.isNotEmpty()) {
            block.append("- warnings: ").append(warningsValue).append("\n")
        }
        return block.toString()
    }

    class QuestAnchorSnapshot(
        templateId: String?,
        questCode: String?,
        questStatus: String?,
        objectiveKey: String?,
        objectiveType: String?,
        reference: String?,
        anchorType: String?,
        anchorId: String?,
        anchorLabel: String?,
        private val updatedAtValue: Long
    ) {
        private val templateIdValue: String = valueOrEmpty(templateId)
        private val questCodeValue: String = valueOrEmpty(questCode)
        private val questStatusValue: String = valueOrEmpty(questStatus)
        private val objectiveKeyValue: String = valueOrEmpty(objectiveKey)
        private val objectiveTypeValue: String = valueOrEmpty(objectiveType)
        private val referenceValue: String = valueOrEmpty(reference)
        private val anchorTypeValue: String = valueOrEmpty(anchorType)
        private val anchorIdValue: String = valueOrEmpty(anchorId)
        private val anchorLabelValue: String = valueOrEmpty(anchorLabel)

        fun templateId(): String = templateIdValue
        fun questCode(): String = questCodeValue
        fun questStatus(): String = questStatusValue
        fun objectiveKey(): String = objectiveKeyValue
        fun objectiveType(): String = objectiveTypeValue
        fun reference(): String = referenceValue
        fun anchorType(): String = anchorTypeValue
        fun anchorId(): String = anchorIdValue
        fun anchorLabel(): String = anchorLabelValue
        fun updatedAt(): Long = updatedAtValue
    }

    companion object {
        @JvmStatic
        fun empty(): StoryContextSnapshot = StoryContextSnapshot(
            "",
            "",
            "",
            WorldContextSnapshot.empty(),
            null,
            null,
            emptyList(),
            emptyList(),
            emptyList(),
            emptyList()
        )

        private fun valueOrEmpty(value: String?): String = value ?: ""

        private fun valueOrUnknown(value: String?): String = if (value.isNullOrBlank()) "unknown" else value
    }
}
